combinetf2/fitter.py

```diff
@@ -0,0 +1,1410 @@
+            )
+
+        self.chisqFit = options.chisqFit
+        self.externalCovariance = options.externalCovariance
+
+        # if quantities are computed with profiling, usually set to true after the fit
+        self.profile = False
+
+        self.nsystgroupsfull = len(self.systgroupsfull)
+
+        self.pois = []
+
+        if options.POIMode == "mu":
+            self.npoi = self.indata.nsignals
+            poidefault = options.POIDefault * tf.ones(
+                [self.npoi], dtype=self.indata.dtype
+            )
+            for signal in self.indata.signals:
+                self.pois.append(signal)
+        elif options.POIMode == "none":
+            self.npoi = 0
+            poidefault = tf.zeros([], dtype=self.indata.dtype)
+        else:
+            raise Exception("unsupported POIMode")
+
+        self.parms = np.concatenate([self.pois, self.indata.systs])
+
+        self.allowNegativePOI = options.allowNegativePOI
+
+        if self.allowNegativePOI:
+            self.xpoidefault = poidefault
+        else:
+            self.xpoidefault = tf.sqrt(poidefault)
+
+        # tf variable containing all fit parameters
+        thetadefault = tf.zeros([self.indata.nsyst], dtype=self.indata.dtype)
+        if self.npoi > 0:
+            xdefault = tf.concat([self.xpoidefault, thetadefault], axis=0)
+        else:
+            xdefault = thetadefault
+
+        self.x = tf.Variable(xdefault, trainable=True, name="x")
+
+        # observed number of events per bin
+        self.nobs = tf.Variable(self.indata.data_obs, trainable=False, name="nobs")
+
+        if self.chisqFit:
+            if self.externalCovariance:
+                if self.indata.data_cov_inv is None:
+                    raise RuntimeError("No external covariance found in input data.")
+                # provided covariance
+                self.data_cov_inv = self.indata.data_cov_inv
+            else:
+                # covariance from data stat
+                if tf.math.reduce_any(self.nobs <= 0).numpy():
+                    raise RuntimeError(
+                        "Bins in 'nobs <= 0' encountered, chi^2 fit can not be performed."
+                    )
+                self.data_cov_inv = tf.linalg.diag(tf.math.reciprocal(self.nobs))
+
+        # constraint minima for nuisance parameters
+        self.theta0 = tf.Variable(
+            tf.zeros([self.indata.nsyst], dtype=self.indata.dtype),
+            trainable=False,
+            name="theta0",
+        )
+
+        # global observables for mc stat uncertainty
+        self.beta0 = tf.Variable(
+            tf.ones_like(self.indata.data_obs), trainable=False, name="beta0"
+        )
+
+        nexpfullcentral = self.expected_yield()
+        self.nexpnom = tf.Variable(nexpfullcentral, trainable=False, name="nexpnom")
+
+        # parameter covariance matrix
+        self.cov = tf.Variable(self.prefit_covariance(), trainable=False, name="cov")
+        self.hess = tf.zeros(
```

Comment:
why is this initialized with zeros?

also, looks like only the top-left corner of this is ever explicitly used (and only after the fit)

Reply:
Removed this. I pass the hessian to the impacts calculation directly now, without storing it permanently